### ISSUE TITLE
If debugging is terminated by an error, debugging may not start until vscode is restarted

### ISSUE THREAD

**Issue by zero-plusplus** [zero-plusplus/vscode-autohotkey-debug#291]
This issue has existed for some time, but could not be fixed due to lack of reproducibility.

I guess that it is probably caused by an error somewhere that is not caught and the termination process is not executed properly.

**Comment by ThioJoe** [comment]
I think i found a reproducable example. Currently I'm running AHK 2.1-Alpha.18, and using UIAccess. Not sure that makes a difference.

I've been having this issue with an object variable created by the UIA library ([available here](https://github.com/Descolada/UIA-v2/blob/90edd2f0ac7895f9caf5bf0ad5b54303c7d4b4fb/Lib/UIA.ahk)).'

This example script should cause it (Note it requires the UIA library to be included). Put the breakpoint at the comment. The weird thing I noticed is that while the `treeViewEl` stops working in the "Watch" panel if added as a watched variable (saying "not available"), it still does seem expandable in the "Local" variables panel, even on subsequent broken re-runs.

```ahk
#Requires AutoHotkey v2.0
#SingleInstance force
#Include "Lib\UIA.ahk"


doTest() {
    MouseGetPos(&mouseX, &mouseY, &winId, &controlHwnd, 3)

    treeViewIdentifier := {ClassName: "SysTreeView32", ControlType: UIA.ControlType.Tree}

    treeViewEl := UIA.ElementFromHandle(controlHwnd)

    if (!treeViewEl) { ; PUT BREAKPOINT HERE
        return
    }
}

; Use this while the mouse is over a SysTreeView, such as the navigation pane in Windows Explorer
F7::doTest()
```

**Comment by zero-plusplus** [comment]
Thank you for the information.
I apologize, but security reasons here prevent us from using third-party scripts.

This issue is probably caused by some exception that prevents the debug adapter from terminating successfully. (i.e., there can be more than one factor).

The bug in the watch panel is probably caused by previous debugging information not being reset.

This extension was created when I had no knowledge of it and it would be very difficult to fix this problem now. I had redesigned and rewritten it from scratch to make it easier to find the problem, but for various reasons I had to stop the project.

Currently, I am devoting development resources to releasing [this extension](https://github.com/zero-plusplus/autohotkey-devtools/tree/prototype/v0.0.1), so fixing this extension has been put on the back burner.

I'm sorry I couldn't do anything about the information you gave me.

It will take some time, but I intend to provide a new debug adapter in the extension linked above, so I will fix this problem there.

**Comment by ThioJoe** [comment]
Totally understandable about third party scripts. And that's fine about the development being focused on the new extension, I figure the info still might come in handy.

I actually just noticed that when the `ECONNRESET` error occurrs, there are windows Event Viewer logs that show up and reveal that Autohotkey itself was killed for an access violation:
```
Faulting application name: AutoHotkey64_UIA.exe, version: 2.1.0.0, time stamp: 0x67e74024
Faulting module name: uiautomationcore.dll, version: 7.2.26100.4061, time stamp: 0x40a5170f
Exception code: 0xc0000005
Fault offset: 0x000000000012a393
Faulting process id: 0xD578
Faulting application start time: 0x1DBC770BFDFF1D5
Faulting application path: C:\Program Files\AutoHotkey\v2\AutoHotkey64_UIA.exe
Faulting module path: C:\Windows\System32\uiautomationcore.dll
Report Id: b5544fbe-4c46-4a87-94fb-eca1107ba7bf
Faulting package full name: 
Faulting package-relative application ID: 
```

So I think the problem occurs when the AutoHotkey process itself is being killed by Windows.

**Comment by zero-plusplus** [comment]
I have reviewed the source code based on your information and the following part may be the problem.

https://github.com/zero-plusplus/vscode-autohotkey-debug/blob/c687a78eb27c2557a7d85e81a5263f0c2570a9df/src/ahkDebug.ts#L194

The close method is implemented as follows.

https://github.com/zero-plusplus/vscode-autohotkey-debug/blob/c687a78eb27c2557a7d85e81a5263f0c2570a9df/src/dbgpSession.ts#L958-L969

Here socketClosed is defined as follows.

https://github.com/zero-plusplus/vscode-autohotkey-debug/blob/c687a78eb27c2557a7d85e81a5263f0c2570a9df/src/dbgpSession.ts#L502-L504

Perhaps the socket fails to recognize that the runtime has terminated due to an external problem and mistakenly believes that it can still write messages to the runtime.

In other words, simply specifying a timeout in the socket termination process may solve this problem.

**Comment by zero-plusplus** [comment]
Fixed in [this branch](https://github.com/zero-plusplus/vscode-autohotkey-debug/tree/refactoring-started-with-2025/01/20).

If you want to try it, clone the above branch locally and execute the following command. This will generate vsix, which you can then use to test with your vscode. **Cloning requires git. Also, Node.js must be installed to execute the following commands.**

* `npm install`
* `npm run package`

The above command will generate `vscode-autohotkey-debug-1.11.1.vsix` in the project folder.

> **Note: The generated vsix cannot be redistributed according to the license, so please use it for personal use.**

---

Since it takes time to officially release a modified version, you may have to wait a few months if you cannot try the above method for security reasons.

**Comment by ThioJoe** [comment]
I finally got around to testing that build, unfortunately it still seems to do the same thing. I'll see if I can do some more testing.

**Comment by zero-plusplus** [comment]
Thanks for the info.

Another possibility is that the following line may be stuck before it is executed.

https://github.com/zero-plusplus/vscode-autohotkey-debug/blob/5961ca8f5eb5a43ea0b7e3fcdab452f081bbd334/src/ahkDebug.ts#L204

Sorry for the hassle, but try the following.

1. Move the above line (line 204) to the first line of the `disconnectRequest` method (line 155)
2. Install as vsix and try it, or debug vscode-autohotkey-debug directly

https://github.com/zero-plusplus/vscode-autohotkey-debug/blob/5961ca8f5eb5a43ea0b7e3fcdab452f081bbd334/src/ahkDebug.ts#L154

---

I will try to find other places where the issue occurs.

**Comment by ThioJoe** [comment]
That doesn't seem to help either unfortunately. I wonder if it might be some kind of state issue with VSCode itself, not necessarily the extension.


I did some more testing and enabled detailed tracing for the extension. This is an example of the output at the time of the crash:

```
2025/05/01 13:04:56.474 Trace <command>property_get -i 199 -n treeViewEl.WinId -c 0 -d 0
2025/05/01 13:04:56.475 Trace <command>property_get -i 200 -n treeViewEl.<enum> -c 0 -d 0
2025/05/01 13:04:58.812 Trace evaluateRequest
2025/05/01 13:04:58.813 Trace <command>context_names -i 201
2025/05/01 13:04:58.813 Trace <command>property_get -i 202 -n treeViewEl -c 0 -d 0
2025/05/01 13:05:05.323 Trace session error
Session closed for the following reasons: read ECONNRESET
2025/05/01 13:05:05.323 Trace sendTerminateEvent
2025/05/01 13:05:05.324 Trace session close
2025/05/01 13:05:05.325 Trace disconnectRequest
Debugging disconnected. AutoHotkey script is continued.
```

Seems like the extension might be correctly recognizing the disconnect.


I thought maybe the socket wasn't being closed or something but when I used TcpLogView it seems like all the connections that are opened are indeed closed when it crashes:

```
| Event Time          | Event Type | Local Address | Remote Address | Local Port | Remote Port | Process ID | Process Name     | Process Path                                                   | Process User 
| ------------------- | ---------- | ------------- | -------------- | ---------- | ----------- | ---------- | ---------------- | -------------------------------------------------------------- | ------------ 
| 6/1/2025 1:04:23 PM | Listen     | 127.0.0.1     | 0.0.0.0        | 9002       | 0           | 40340      | Code.exe         | C:\Users\Joe\AppData\Local\Programs\Microsoft VS Code\Code.exe | JOE-PC\Joe   
| 6/1/2025 1:04:23 PM | Open       | 127.0.0.1     | 127.0.0.1      | 9002       | 54042       | 40340      | Code.exe         | C:\Users\Joe\AppData\Local\Programs\Microsoft VS Code\Code.exe | JOE-PC\Joe   
| 6/1/2025 1:04:23 PM | Open       | 127.0.0.1     | 127.0.0.1      | 54042      | 9002        | 48428      | AutoHotkey64.exe | C:\Program Files\AutoHotkey\v2\AutoHotkey64.exe                | JOE-PC\Joe   
| 6/1/2025 1:05:03 PM | Open       | 192.168.1.103 | 13.89.179.12   | 54077      | 443         | 4360       | WerFault.exe     | C:\Windows\System32\WerFault.exe                               | JOE-PC\Joe   
| 6/1/2025 1:05:04 PM | Close      | 192.168.1.103 | 13.89.179.12   | 54077      | 443         | 4360       | WerFault.exe     | C:\Windows\System32\WerFault.exe                               | JOE-PC\Joe   
| 6/1/2025 1:05:05 PM | Close      | 127.0.0.1     | 0.0.0.0        | 9002       | 0           | 40340      | Code.exe         | C:\Users\Joe\AppData\Local\Programs\Microsoft VS Code\Code.exe | JOE-PC\Joe   
| 6/1/2025 1:05:05 PM | Close      | 127.0.0.1     | 127.0.0.1      | 9002       | 54042       | 40340      | Code.exe         | C:\Users\Joe\AppData\Local\Programs\Microsoft VS Code\Code.exe | JOE-PC\Joe   
| 6/1/2025 1:05:05 PM | Close      | 127.0.0.1     | 127.0.0.1      | 54042      | 9002        | 48428      | AutoHotkey64.exe | C:\Program Files\AutoHotkey\v2\AutoHotkey64.exe                | JOE-PC\Joe   
```

This is the trace of starting a debug session where VSCode is in the 'bad' state:
```
2025/05/01 12:41:39.385 Trace launchRequest
"C:\Program Files\AutoHotkey\v2\AutoHotkey64.exe" /Debug=127.0.0.1:9002 /ErrorStdOut B:\Dropbox\Backup\Program Settings and Configurations\Autohotkey Scripts\Testing\TestCrash.ahk
2025/05/01 12:41:39.454 Trace <command>feature_set -i 2 -n max_depth -v 1
2025/05/01 12:41:39.454 Trace <command>stdout -i 3 -c 2
2025/05/01 12:41:39.454 Trace <command>stderr -i 4 -c 2
2025/05/01 12:41:39.454 Trace <command>property_set -i 5 -n A_DebuggerName -c 1 -- VmlzdWFsIFN0dWRpbyBDb2Rl
2025/05/01 12:41:39.454 Trace <command>feature_get -i 6 -n language_version
Debugger Adapter Version: 1.11.1
Debug Configuration (launch): AutoHotkey Debug (Max Output)
AutoHotkey Version: 2.1-alpha.18
2025/05/01 12:41:39.466 Trace setBreakPointsRequest
2025/05/01 12:41:39.467 Trace <command>breakpoint_set -i 7 -t line -f file:///b%3A/Dropbox/Backup/Program%20Settings%20and%20Configurations/Autohotkey%20Scripts/Testing/TestCrash.ahk -n 11
2025/05/01 12:41:39.467 Trace <command>breakpoint_get -i 8 -d 1
2025/05/01 12:41:39.467 Trace <command>breakpoint_set -i 9 -t line -f file:///b%3A/Dropbox/Backup/Program%20Settings%20and%20Configurations/Autohotkey%20Scripts/Testing/TestCrash.ahk -n 15
2025/05/01 12:41:39.468 Trace <command>breakpoint_get -i 10 -d 2
2025/05/01 12:41:39.469 Trace configurationDoneRequest
2025/05/01 12:41:39.469 Trace <command>feature_set -i 11 -n max_children -v 10000
2025/05/01 12:41:39.471 Trace <command>run -i 12
2025/05/01 12:41:39.471 Trace threadsRequest
2025/05/01 12:41:45.381 Trace checkContinuationStatus
2025/05/01 12:41:45.381 Trace <command>stack_get -i 13
2025/05/01 12:41:45.382 Trace sendStoppedEvent
2025/05/01 12:41:45.391 Trace threadsRequest
2025/05/01 12:41:45.392 Trace stackTraceRequest
2025/05/01 12:41:45.797 Trace scopesRequest
2025/05/01 12:41:45.797 Trace <command>context_names -i 14
2025/05/01 12:41:45.799 Trace variablesRequest
2025/05/01 12:41:45.799 Trace <command>context_get -i 15 -c 0 -d 0
```

And this is doing the same thing but after restarting VSCode:
```
2025/05/01 12:45:11.349 Trace launchRequest
"C:\Program Files\AutoHotkey\v2\AutoHotkey64.exe" /Debug=127.0.0.1:9002 /ErrorStdOut B:\Dropbox\Backup\Program Settings and Configurations\Autohotkey Scripts\Testing\TestCrash.ahk
2025/05/01 12:45:11.417 Trace <command>feature_set -i 2 -n max_depth -v 1
2025/05/01 12:45:11.417 Trace <command>stdout -i 3 -c 2
2025/05/01 12:45:11.417 Trace <command>stderr -i 4 -c 2
2025/05/01 12:45:11.417 Trace <command>property_set -i 5 -n A_DebuggerName -c 1 -- VmlzdWFsIFN0dWRpbyBDb2Rl
2025/05/01 12:45:11.418 Trace <command>feature_get -i 6 -n language_version
Debugger Adapter Version: 1.11.1
Debug Configuration (launch): AutoHotkey Debug (Max Output)
AutoHotkey Version: 2.1-alpha.18
2025/05/01 12:45:11.454 Trace setBreakPointsRequest
2025/05/01 12:45:11.455 Trace <command>breakpoint_set -i 7 -t line -f file:///b%3A/Dropbox/Backup/Program%20Settings%20and%20Configurations/Autohotkey%20Scripts/Testing/TestCrash.ahk -n 11
2025/05/01 12:45:11.456 Trace <command>breakpoint_get -i 8 -d 1
2025/05/01 12:45:11.456 Trace <command>breakpoint_set -i 9 -t line -f file:///b%3A/Dropbox/Backup/Program%20Settings%20and%20Configurations/Autohotkey%20Scripts/Testing/TestCrash.ahk -n 15
2025/05/01 12:45:11.457 Trace <command>breakpoint_get -i 10 -d 2
2025/05/01 12:45:11.459 Trace configurationDoneRequest
2025/05/01 12:45:11.460 Trace <command>feature_set -i 11 -n max_children -v 10000
2025/05/01 12:45:11.462 Trace <command>run -i 12
2025/05/01 12:45:11.462 Trace threadsRequest
2025/05/01 12:45:15.459 Trace checkContinuationStatus
2025/05/01 12:45:15.460 Trace <command>stack_get -i 13
2025/05/01 12:45:15.461 Trace sendStoppedEvent
2025/05/01 12:45:15.471 Trace threadsRequest
2025/05/01 12:45:15.472 Trace stackTraceRequest
2025/05/01 12:45:15.537 Trace evaluateRequest
2025/05/01 12:45:15.538 Trace <command>context_names -i 14
2025/05/01 12:45:15.539 Trace <command>property_get -i 15 -n controlHwnd -c 0 -d 0
2025/05/01 12:45:15.540 Trace evaluateRequest
2025/05/01 12:45:15.540 Trace <command>context_names -i 16
2025/05/01 12:45:15.540 Trace <command>property_get -i 17 -n controlHwnd -c 0 -d 0
2025/05/01 12:45:15.880 Trace scopesRequest
2025/05/01 12:45:15.880 Trace <command>context_names -i 18
2025/05/01 12:45:15.882 Trace variablesRequest
2025/05/01 12:45:15.882 Trace <command>context_get -i 19 -c 0 -d 0
```

The difference is the bad state one is missing some lines for `evaluateRequest` and `<command>property_get`. Those would be for variables that are already in the watch panel. But also nothing happens if I try to add a new watch variable, it just doesn't show any trace things happening and no new variables get added to the panel.

------------

Side Note:

I've also noticed that while other methods of forcefully terminating the AutoHotkey process (like Task Manager, pskill, or even a debugger forced crash) can trigger an ECONNRESET error, the ECONNRESET by itself doesn't consistently lead to the issue of watched variables not being available. That's only reliably reproduced when the AutoHotkey process crashes due to an access violation (0xc0000005) where uiautomationcore.dll is specifically identified as the faulting module, rather than a crash originating directly within the main AutoHotkey executable or a generic termination.

Perhaps another hint that maybe the issue is with VSCode not handling something and not that the extension isn't handling the disconnect.

**Comment by zero-plusplus** [comment]
Thanks for the additional info.
If the same issue occurs with other debug adapters, we can see if it is an extension-specific issue.

But, other debug adapters may not be able to debug the UIA version of the runtime (At least with this debug adapter, a workaround was necessary).
